### PR TITLE
Refactor report generation tool

### PR DESCRIPTION
**Pull request by erfanio**
This is the first step towards making reports directly read/write to both local disk and GCS.

### CODE REVIEW

**Comment by erfanio** [comment]
Chatted with Dongge in person. Getting rid of the stacked PRs because they're such a pain to maintain, but keeping the commit history clean to make it easy to review commit by commit.

**Comment by DonggeLiu** [comment]
The report looks good!
https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-06-13-328-dg-comparison/index.html

**Comment by erfanio** [comment]
Just rebased on main again. Should be good to merge.

**Comment by erfanio** [comment]
Hmmm https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-06-14-328-dg-comparison/index.html has a bunch of benchmarks still running even though the experiment has finished

**Comment by DonggeLiu** [comment]
> Hmmm https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-06-14-328-dg-comparison/index.html has a bunch of benchmarks still running even though the experiment has finished

Did you refresh? : )

**Comment by erfanio** [comment]
Interesting... the [experiment node](https://pantheon.corp.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-328-dg/details?e=-13802955&mods=logs_tg_prod&project=oss-fuzz) has finished though, so the report will stay the same even after they finish. I thought the experiment would block on GCB tasks finishing

**Comment by DonggeLiu** [comment]
> Interesting... the [experiment node](https://pantheon.corp.google.com/kubernetes/job/us-central1-c/llm-experiment/default/ofg-pr-328-dg/details?e=-13802955&mods=logs_tg_prod&project=oss-fuzz) has finished though, so the report will stay the same even after they finish. I thought the experiment would block on GCB tasks finishing

Not sure if I missed anything, but I think it is still running?
image

**Comment by erfanio** [comment]
Oh yeah :facepalm: I'm blind haha

**Review by DonggeLiu — APPROVED** [review]
Thanks! @erfanio

File: report/web.py [hunk 1]
```diff
@@ -135,363 +89,441 @@ class Target:
...
+        filter(self._is_valid_benchmark_dir, os.listdir(self._results_dir)))
 
-  results = []
-  status_dir = os.path.join(RESULTS_DIR, benchmark, 'status')
+  def list_benchmarks(self) -> List[Benchmark]:
+    """Lists benchmarks in the result directory."""
+    benchmarks = []
+    for benchmark in self.list_benchmark_ids():
+      results, targets = self._get_results(benchmark)
+      status = 'Done' if all(r for r in results) and results else 'Running'
```
[the remark this answers is not in the document]
Comment (erfanio) [reply report/web.py:108]: I'm pretty much re-writing large parts of these functions in the next PR so let's make the changes then.

**Pull request merged**